### PR TITLE
ENH: Add vtkPlusCommand for Capistrano device

### PR DESCRIPTION
**Pull request by nhjohnston**
## 1. Add functions (9) for CapistranoDevice Remote Control under the command **CapistranoCommand**:

 **FreezeProbe, IsProbeFrozen​
 BidirectionalMode​
 SetJitterCompensation, GetJitterCompensation​
 SetMISMode, SetMISPulsePeriod​
 SetWobbleRate, GetWobbleRate**

With XML strings:
##### FreezeProbe (bool)
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="FreezeProbe" Value="1" /></Command> `
##### BidirectionalMode​ (bool)
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="SetBidirectionalMode" Value="1" /></Command> `
##### SetJitterCompensation
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="SetJitterCompensation" Value="40" /></Command>`
##### SetMISMode (bool)
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="SetMISMode" Value="0" /></Command> `
##### SetMISPulsePeriod
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="SetMISPulsePeriod" Value="100" /></Command> `
##### SetWobbleRate
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="SetWobbleRate" Value="100" /></Command>`
##### IsProbeFrozen
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="IsProbeFrozen" /></Command>  `
##### GetJitterCompensation
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="GetJitterCompensation" /></Command> `
##### GetWobbleRate (to check the functionality of SetWobbleRate)
`<Command Name="CapistranoCommand" UsDeviceId="Wobbler"><Parameter Name="GetWobbleRate" /></Command> `

## 2. Updated the CapistranoVideoSource
Updated the CapistranoVideoSource `InternalApplyImagingParameterChange` function to use `SetGainPercentDevice` function rather than the `SetGainPercent` function, as the original function had no effect on the gain of the device.

### CODE REVIEW

File: src/PlusDataCollection/Capistrano/vtkPlusCapistranoVideoSource.cxx [hunk 1]
```diff
@@ -2047,7 +2060,7 @@ PlusStatus vtkPlusCapistranoVideoSource::InternalApplyImagingParameterChange()
     this->ImagingParameters->GetTimeGainCompensation(tgcVec);
     double tgc[3] = { tgcVec[0], tgcVec[1], tgcVec[2] };
 
-    if (this->SetGainPercent(tgc) == PLUS_SUCCESS)
+    if (this->SetGainPercentDevice(tgc) == PLUS_SUCCESS)
```
Comment (adamaji) [inline src/PlusDataCollection/Capistrano/vtkPlusCapistranoVideoSource.cxx:2063]: this should probably go in a different commit as it's somewhat separate from the command addition? ("BUG: Capistrano TGC not updated through remote commands" or something like that)

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+}
+
+//----------------------------------------------------------------------------
+std::string vtkPlusCapistranoCommand::GetDescription(const std::string& commandName)
+{
+  std::string desc;
+  if (commandName.empty() || igsioCommon::IsEqualInsensitive(commandName, CAPISTRANO_COMMAND))
+  {
+    desc += CAPISTRANO_COMMAND;
+    desc += ": Send text data to the device.";
```
Comment (adamaji) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: the description should probably match the device class. maybe "Send commands to the Capistrano device."?

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+      ss << value;
+      resultString += " Success=\"true\"";
+      resultString += " Value=\"" + ss.str() + "\"";
+      metaData[parameterName] = std::make_pair(IANA_TYPE_US_ASCII, ss.str());
+      status = PLUS_SUCCESS;
+    }
+    if (parameterName == vtkPlusCapistranoVideoSource::CAPISTRANO_SET_MIS_MODE)
+    {
+      bool MISMode_bool = std::stoi(value);
+      status = device->SetMISMode(MISMode_bool);
```
Comment (adamaji) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: Though it probably won't happen in practice since we generally build with the correct sdk, this function (and a few others) have a chance of failure (https://github.com/PlusToolkit/PlusLib/blob/62ab3159e91eda69b2b19ffd22e46bde74590997/src/PlusDataCollection/Capistrano/vtkPlusCapistranoVideoSource.cxx#L1471). 

SetUsParameter handles the case of commands failing to actually do the thing they're supposed to by relying on the output of `SetNewImagingParameters`/`InternalApplyImagingParameterChanges`. 

Is there something similar we can do here to ensure that the success/failure status return sent back by the command reflects the actual state of the device post-command-being-sent?

**Review by adamaji — COMMENTED** [review]
I would change the main commit message here to something like "ENH: Add vtkPlusCommand for Capistrano device", since technically the Capistrano device should be able to be controlled through SetUsParameter already

File: src/PlusServer/vtkPlusCommandProcessor.cxx [hunk 4]
```diff
@@ -38,6 +38,7 @@ See License.txt for details.
 #include "vtkPlusStartStopRecordingCommand.h"
 #include "vtkPlusUpdateTransformCommand.h"
 #include "vtkPlusVersionCommand.h"
+#include "vtkPlusCapistranoCommand.h"
```
Comment (jamesobutler) [inline src/PlusServer/vtkPlusCommandProcessor.cxx]: When building the [PlusBuild](https://github.com/PlusToolkit/PlusBuild) project pointing to this PlusLib branch and WITHOUT Capistrano device support there are linking errors and the project fails to build successfully. This include of `vtkPlusCapistranoCommand.h` likely causes this error because the `vtkPlusCapistranoCommand` class contains  `#include "vtkPlusCapistranoVideoSource.h"`
```
8>  vtkPlusCapistranoCommand.cxx
8>C:\P\PR22\PlusLib\src\PlusServer\Commands\vtkPlusCapistranoCommand.cxx(10,10): fatal error C1083: Cannot open include file: 'vtkPlusCapistranoVideoSource.h': No such file or directory [C:\P\PR22\PlusLib-bin\src\PlusServer\vtkPlusServer.vcxproj]
```

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+#include "vtkPlusCommand.h"
+#include "vtkIGSIOTransformRepository.h"
+
+class vtkMatrix4x4;
+
+/*!
+  \class vtkPlusCapistranoCommand
+  \brief This command is for communicating with vtkPlusCapistranoSerialDevices.
+  \ingroup PlusLibPlusServer
+  This command is used for communicating with a Capistrano serial device.
```
Comment (jamesobutler) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.h]: There seem to be some copy/paste stuff that isn't correct such as describing "Capistrano" as a serial device. Some comments here and the `SetNameToSerial` aren't valid as it is not a serial device. This appears to have been copied over from the GenericSerialDevice class.

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx (hunk 3 above)
Comment (nhjohnston) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: Like checking the status of the command and adding a log and plus status change?
(for commands: SetMISMode, SetMISPlusePeriod, FreezeDevice) 
EX: https://github.com/nhjohnston/PlusLib/blob/35e8d8eea6c6faa043c65d6cc2590a74dcca512f/src/PlusDataCollection/Capistrano/vtkPlusCapistranoVideoSource.cxx#L2123-L2132

Comment (adamaji) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: Yes, something like that should work

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    if (parameterName == vtkPlusCapistranoVideoSource::CAPISTRANO_SET_BIDIRECTIONAL_MODE)
+    {
+      bool bidi = std::stoi(value);
+      status = device->SetBidirectionalMode(bidi);
+      resultString += " Success=\"true\"/>";
+      metaData[parameterName] = std::make_pair(IANA_TYPE_US_ASCII, "SUCCESS");
+    }
+    if (parameterName == vtkPlusCapistranoVideoSource::CAPISTRANO_FREEZE_PROBE)
+    {
+      bool freezeProbe_bool  = std::stoi(value);
```
Comment (adamaji) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: it's a little weird that the setter for the frozen parameter takes as input ("0"/"1") while the getter returns ("true"/"false"). we should probably be consistent here?

additionally, if you try to send a string like "true" through this command with these changes, the server does crash. do you think it's necessary to return parse failures like in https://github.com/PlusToolkit/PlusLib/blob/9735b2b70e7db9996dd7fb8534e85b58fb90985c/src/PlusServer/Commands/vtkPlusSetUsParameterCommand.cxx#L220 ?

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+    vtkPlusCapistranoVideoSource* device = dynamic_cast<vtkPlusCapistranoVideoSource*>(aDevice);
+
+    if (parameterName == vtkPlusCapistranoVideoSource::CAPISTRANO_SET_WOBBLE_RATE)
+    {
+      int wobbleRate_int = std::stoi(value);
+      status = device->SetWobbleRate((unsigned char)wobbleRate_int);
+      resultString += " Success=\"true\"/>";
+      metaData[parameterName] = std::make_pair(IANA_TYPE_US_ASCII, "SUCCESS");
+    }
+    if (parameterName == vtkPlusCapistranoVideoSource::CAPISTRANO_GET_WOBBLE_RATE)
```
Comment (adamaji) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: these can be `else if`s since `parameterName` is only ever one string at a time

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+      ss << value;
+      resultString += " Success=\"true\"";
+      resultString += " Value=\"" + ss.str() + "\"";
+      metaData[parameterName] = std::make_pair(IANA_TYPE_US_ASCII, ss.str());
+      status = PLUS_SUCCESS;
+    }
+    if (parameterName == vtkPlusCapistranoVideoSource::CAPISTRANO_SET_MIS_MODE)
+    {
+      bool MISMode_bool = std::stoi(value);
+      if (device->SetMISMode(MISMode_bool)== PLUS_SUCCESS)
```
Comment (adamaji) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: not sure if the a-formatting will catch this, but needs spaces between ` == `

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_GET_WOBBLE_RATE
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_SET_MIS_MODE
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_SET_MIS_PULSE_PERIOD
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_GET_JITTER_COMPENSATION
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_SET_JITTER_COMPENSATION
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_FREEZE_PROBE
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_IS_PROBE_FROZEN
+        && parameterName != vtkPlusCapistranoVideoSource::CAPISTRANO_SET_BIDIRECTIONAL_MODE)
+    {
+      error += "Invalid parameter " + parameterName + ". ";
```
Comment (adamaji) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx:224]: For some reason these error messages don't get propagated to the CommandReply? This is despite the later line `commandResponse->SetErrorString(error);`. I'm not sure why this is, SetUsParameter seems to have similar behavior

image

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx (hunk 6 above)
Comment (nhjohnston) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: EDIT: Will use true/false to align with WinprobeCommands

Is there a preference for using true/false vs 0/1? Using 0/1 makes checking for parsing errors easier - like the UsParameters example. True/False may be more intuitive for a user, but the returns will most often be handled by SonoEQ, so I am not sure how important that is

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx (hunk 9 above)
Comment (nhjohnston) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx:224]: I think the error message shows up in the second response window (for both Set/GetUsParameter and CapsitranoCommand). Is this not as expected?
image

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx (hunk 6 above)
Comment (adamaji) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: I think 0/1 makes sense. @jamesobutler do you have a preference here?

Comment (adamaji) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: Sorry didn't see the edit here. Carry on

Comment (nhjohnston) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: > I think 0/1 makes sense. @jamesobutler do you have a preference here?

I would agree, I feel like the code is cleaner with 0/1, but I then wasn't sure about continuity.

If there is no preference I'm going to go ahead with 0/1 to avoid adding unnecessary lines of code

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+      resultString += " Success=\"false\"/>";
+      metaData[parameterName] = std::make_pair(IANA_TYPE_US_ASCII, "FAIL");
+      status = PLUS_FAIL;
+      continue;
+    } 
+    vtkPlusCapistranoVideoSource* device = dynamic_cast<vtkPlusCapistranoVideoSource*>(aDevice);
```
Comment (jamesobutler) [inline src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx]: This device cast should likely be outside of the `for` loop.

File: src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx (hunk 9 above)
Comment (adamaji) [reply src/PlusServer/Commands/vtkPlusCapistranoCommand.cxx:224]: Oh I see, I was confused as there are some error responses from the GenericSerial command class that get passed through the CommandReply XML string. I suppose this is fine.
